### PR TITLE
Expose types from `sc-service`

### PR DESCRIPTION
**Pull request by RomarQ**
# Description

At moonbeam we have worked on a `lazy-loading` feature which is a client mode that forks a live parachain and fetches its state on-demand, we have been able to do this by duplicating some code from `sc_service::client`. The objective of this PR is to simplify the implementation by making public some types in polkadot-sdk.

- Modules:
   -  `sc_service::client` **I do not see a point to only expose this type when `test-helpers` feature is enabled**

## Integration

Not applicable, the PR just makes some types public.

## Review Notes

The changes included in this PR give more flexibility for client developers by exposing important types.

### CODE REVIEW

**Comment by bkchr** [comment]
> * `sc_service::client` **I do not see a point to only expose this type when `test-helpers` feature is enabled**

This feature should not exist at all...

Could you give some more details on how you implemented this? I mean why do you need to expose these types at all? I would say that you only need to write some `Backend` that fetches the values on demand (for sure it would not be async and whatever, but we don't support this any way right now).

**Comment by RomarQ** [comment]
> > * `sc_service::client` **I do not see a point to only expose this type when `test-helpers` feature is enabled**
> 
> This feature should not exist at all...
> 
> Could you give some more details on how you implemented this? I mean why do you need to expose these types at all? I would say that you only need to write some `Backend` that fetches the values on demand (for sure it would not be async and whatever, but we don't support this any way right now).

We had to change the `CallExecutor` because of some validation in `contextual_call`, but there was no way to pass it a `Client` instance without the use of `test-helpers` feature.

**Comment by bkchr** [comment]
> `` because of some validation in `contextual_call`

Which validation? Could you be more specific?

**Comment by RomarQ** [comment]
> > `` because of some validation in `contextual_call`
> 
> Which validation? Could you be more specific?

It was the [as_trie_backend](https://github.com/paritytech/polkadot-sdk/blob/06a68bec30dd56395fb8ae03e80e0fdadc682ed9/substrate/client/service/src/client/call_executor.rs#L154C4-L173C5) when recording is enabled, we cannot convert the lazy loading to a `trie backend` since we only have access to the state on-demand.

**Comment by bkchr** [comment]
Dumb question, could you not just use a custom `CallExecutor` that wraps the `LocalCallExecutor` and always forwards `recorder` as `None`?

**Comment by RomarQ** [comment]
> Dumb question, could you not just use a custom `CallExecutor` that wraps the `LocalCallExecutor` and always forwards `recorder` as `None`?

The `Client` struct is still not public without the use of `test-helpers` feature. About wrapping `LocalCallExecutor`, it is for sure possible, I have updated the PR.

**Comment by bkchr** [comment]
Can you please remove the `test-helpers` feature completely?

**Comment by RomarQ** [comment]
> Can you please remove the `test-helpers` feature completely?

Sure, will have a look at it :+1:

**Review by bkchr — APPROVED** [review]

File: substrate/client/service/src/lib.rs [hunk 1]
```diff
@@ -27,10 +27,7 @@ pub mod config;
 pub mod error;
 
 mod builder;
-#[cfg(feature = "test-helpers")]
 pub mod client;
```
Comment (michalkucharczyk) [inline substrate/client/service/src/lib.rs:30]: nit: could be grouped with other pub mods.

**Review by michalkucharczyk — APPROVED** [review]

File: substrate/client/service/src/lib.rs (hunk 1 above)
Comment (RomarQ) [reply substrate/client/service/src/lib.rs]: Done :+1:

**Comment by RomarQ** [comment]
> /cmd prdoc --bump major --audience node_dev

added the PRdoc

**Comment by RomarQ** [comment]
@bkchr can this PR be merged?

**Pull request merged**